doc/regexp.rdoc

```diff
@@ -1228,6 +1228,34 @@ when regexp.timeout is non-+nil+, that value controls timing out:
   |         nil          |         Float        | Times out in Float seconds. |
   |        Float         |          Any         | Times out in Float seconds. |
 
+== Optimization
+
+Regexp matching time may increase explosively depending on a pattern and an
+target string. A vulnerability caused by this is known as {regular expression
+denial-of-service (ReDoS)}[https://en.wikipedia.org/wiki/ReDoS].
+
+Ruby's regexp matching applies an optimization to prevent ReDoS. When the
+optimization is applied, matching time increases in linear to the input's
+length; that is ReDoS does not cause.  For applying this optimization, a
+pattern satisfy the following conditions:
+
+- no backreferences
+- no subexpression calls
+- no nested lookaround anchors and atomic groups
+- no nested quantifiers with counting (i.e. no nested <tt>{n}</tt>,
+  <tt>{min,}</tt>, <tt>{,max}</tt>, or <tt>{min,max}</tt> style quantifiers)
```

Comment:
\Regexp matching can apply an optimization to prevent ReDoS attacks.
When the optimization is applied, matching time increases linearly (not polynomially or exponentially)
in relation to the input size, and a ReDoS attach is not possible.

This optimization is applied if and only if the pattern meets these criteria:

- No backreferences.
- No subexpression calls.
- No nested lookaround anchors or atomic groups.
- No nested quantifiers with counting (i.e. no nested <tt>{n}</tt>,
  <tt>{min,}</tt>, <tt>{,max}</tt>, or <tt>{min,max}</tt> style quantifiers).

Reply:
"if and only if" sounds too strict. Actually, since some small counting repetition is unfolded internally, this case does not violate the conditions. The complete condition is too complicated to explain and detailed on the implementation. "if" seems better to me.